File: C++/Leetcode 953.cpp
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

class Solution {
    bool Helper(string A, string B,vector<int> &temp)
    {
        int len = min(A.length(),B.length());

        for (int i = 0; i < len; i++)
        {
            if (temp[A[i] - 'a'] <  temp[B[i] - 'a'])
            {
                return true;
            }
            else if (temp[A[i] - 'a'] >  temp[B[i] - 'a'])
            {
                return false;
            }
        }

        if (A.length() > B.length())
        {
            return false;
        }
        return true;
    }
public:
    bool isAlienSorted(vector<string>& words, string order) {
        vector<int> Dictionary(26,0);

        for (int i = 0; i < order.length(); i++)
        {
            Dictionary[order[i] - 'a'] = i;
        }

        for (int i = 0; i < words.size()-1; i++)
        {
            if (!Helper(words[i],words[i+1],Dictionary))
            {
                return false;
            }
        }
        return true;
    }
};

File: Python/Leetcode 953.py
class Solution:
    def Helper(self, str1: str, str2: str, Dict : List) -> bool:
        min_length = min(len(str1),len(str2))

        for index in range(0,min_length):
            if Dict[ord(str1[index]) - ord('a')] < Dict[ord(str2[index]) - ord('a')]:
                return True
            
            elif Dict[ord(str1[index]) - ord('a')] > Dict[ord(str2[index]) - ord('a')]:
                return False
        
        if (len(str1) > len(str2)):
            return False
        
        return True

    def isAlienSorted(self, words: List[str], order: str) -> bool:
        Dict = []

        for i in range(0,26):
            Dict.append(0)

        for i in range(0,len(order)):
            # print(i)
            Dict[ord(order[i]) - ord('a')] = i
        
        for i in range(0,len(words)-1):

            print(i)
            if not self.Helper(words[i],words[i+1],Dict):
                return False
            
        return True
